torchvision/datasets/imagenet.py

```diff
@@ -0,0 +1,197 @@
+    # FIXME: I don't know if this is good pratice / robust
+    FileExistsError = OSError
+
+import shutil
+import scipy.io as sio
+import torch
+from .folder import ImageFolder
+from .utils import check_integrity, download_url
+
+ARCHIVE_DICT = {
+    ('2012', 'train'): {
+        'url': 'http://www.image-net.org/challenges/LSVRC/2012/nnoupb/ILSVRC2012_img_train.tar',
+        'md5': '1d675b47d978889d74fa0da5fadfb00e',
+    },
+    ('2012', 'val'): {
+        'url': 'http://www.image-net.org/challenges/LSVRC/2012/nnoupb/ILSVRC2012_img_val.tar',
+        'md5': '29b22e2961454d5413ddabcf34fc5622',
+    },
+    ('2012', 'devkit'): {
+        'url': 'http://www.image-net.org/challenges/LSVRC/2012/nnoupb/ILSVRC2012_devkit_t12.tar.gz',
+        'md5': 'fa75699e90414af021442c21a62c3abf',
+    }
+}
+
+META_DICT = {
+    '2012': '5c2648af14b2ff44540504b860a81a79',
+}
+
+META_FILE = 'meta.bin'
+
+
+class ImageNetClassification(ImageFolder):
+    def __init__(self, root, split='val', year='2012', download=False, **kwargs):
+
+        root = self.root = os.path.expanduser(root)
+        self.split = self._verify_split(split)
+        self.year = self._verify_year(year)
+
+        if download:
+            self.download()
+
+        self.wnids, self.wnid_to_idx, classes, class_to_idx = self._load_meta()
+        super(ImageNetClassification, self).__init__(self.split_folder, **kwargs)
+        self.root = root
+        self.classes = classes
+        self.class_to_idx = class_to_idx
+
+    def download(self):
+        self._prepare_tree()
+
+        meta_file = os.path.join(self.year_folder, META_FILE)
+        if not check_integrity(meta_file, META_DICT[self.year]):
+            tmpdir = os.path.join(self.root, 'tmp')
+
+            archive_dict = ARCHIVE_DICT[(self.year, 'devkit')]
+            download_and_extract_tar(archive_dict['url'], self.root,
+                                     extract_root=tmpdir,
+                                     md5=archive_dict['md5'], gzip=True)
+            devkit_folder = _splitexts(os.path.basename(archive_dict['url']))[0]
+            meta, val_wnids = parse_devkit(os.path.join(tmpdir, devkit_folder))
+            torch.save((meta, val_wnids), meta_file)
+
+            shutil.rmtree(tmpdir)
+
+        archive_dict = ARCHIVE_DICT[(self.year, self.split)]
+        download_and_extract_tar(archive_dict['url'], self.root,
+                                 extract_root=self.split_folder,
+                                 md5=archive_dict['md5'], gzip=False)
+
+        if self.split == 'val':
+            val_wnids = torch.load(meta_file)[1]
+            prepare_val_folder(self.split_folder, val_wnids)
+
+    def _load_meta(self):
+        # TODO: verify meta file
+        return torch.load(os.path.join(self.year_folder, META_FILE))[0]
+
+    def _prepare_tree(self):
+        try:
+            os.makedirs(self.split_folder)
+        except FileExistsError:
+            shutil.rmtree(self.split_folder)
```

Comment:
Are you sure we want to delete everything and re-download everything if we pass `download=True` again? For me we would just skip re-downloading it

Reply:
I thought about that a bit. I intentionally defaulted the `download` flag to `False` so the user has to explicitly set it. If he does I don't see an option to check if the dataset is corrupted or not. In this case I check if the archive is available and download it if not. The code above only empties the `split_folder` which is for example `ILSVRC/2012/train`. This will be subsequently filled again by extracting the archives.

Reply:
The other datasets do not re-download the files if it's already available. I think this is particularly important for ImageNet, given it's size.

If the dataset is corrupted by some means, the user can himself delete the files and re-download maybe?

Reply:
My implementation also does not re-download the archives, if they are already present. Within the `download_and_extract_tar` the following lines achieve this:
```python
if not check_integrity(os.path.join(download_root, filename), md5):
    download_url(url, download_root, filename=filename, md5=md5)
```
Thus, it is only re-downloaded if
1. the archive is not present in the given root directory, or
2. its MD5 hash does not check out.

What I always do (if the user set `download=True`) is, deleting the contents of the `train` or `val` folder and re-extracting the archive. I do this, because without further information I'm not aware of a method to check the integrity of the extracted dataset.

With the development kit however (see discussion above) we could do some integrity checks:
1. Assert that `train` and `val` folder contain exactly `1000` subfolders with the correct wordnet IDs, and
2. assert that each subfolder contains the correct amount of images.

What do you think?

Reply:
Oh I see, thanks for the explanation!

This looks better to me (compared to re-downloading every time), but the side effect of removing the folders might be surprising.

Imagine for example that someone deleted the zips after the extraction was done, in order to save memory. Then, you would `rm` the whole `train` / `val` folders, requiring the user to re-download everything.

While I think that having more integrity checks is definitely a good thing, I'd rather do a single integrity check on the zips once its downloaded, than re-extract every time.

Thoughts?

Reply:
I need some further clarification: How would your suggestion

> I'd rather do a single integrity check on the zips once its downloaded, than re-extract every time

solve the issue you described above? Without integrity checks of the extracted dataset, we have to re-download, if the user deletes the archives **and** sets `download=True`. Since I defaulted `download=False` this is a deliberate choice by the user and I see no way around  it.

Right now the code  does the following if the user sets `download=True`:
1. Empty the `split_folder`, i. e. delete all `train` or `val` images.
2. Check if the corresponding archive is available. If not, download it.
3. Extract the archive in the (now empty) `split_folder`.

The archive itself is **not** located inside the `split_folder` (take a look at the updated tree in the initial comment) and thus not deleted. IMO this is what you describe.

Reply:
If the user sets `download=True` and there is a folder existing with the expected name, then nothing would be downloaded. One other approach could be to store a dummy file, like `.torchvision_verified` in the root of the extracted dataset, so that it can be used by the verification logic if the zip file is not available.

Thoughts?

Reply:
Only addressing the first part of your comment here. For the second part see my answer below.

I think we are still talking past each other. At no point the code checks if a specific folder is present. I always _empty_ (recursively delete and create again) the `split_folder`. Thus, it doesn't matter whether the  `split_folder` exist or doesn't if the user sets `download=True`.

Reply:
I still think that we should **not** delete the folder at first.

I think it might be more common than what you might imagine that people will go and delete the tar file containing the zipped images. They will be very surprised once they realize that they have to re-download everything because they deleted it.

Here is my point: downloading ImageNet is very time-consuming. If we want to perform some integrity check, we should do it just after downloading it. After we download it and the checks match, we assume that the zip extraction will work fine, and we don't need to worry about checking the hash all the time.

Reply:
> I still think that we should **not** delete the folder **at first**. (additional emphasis added)

Is it about the order of the operation, as in check if we have the archive before we empty the folder? If so, we could simply move the `_empty_split_folder()` call just before the extraction. This makes more sense now, since before the function created the whole tree for the `year`s etc.

> we don't need to worry about checking the hash all the time.

All this checking is only done if the user explicitly sets `download=True` which normally defaults to `False`. If I understand you correctly, all these issues can only occur if the dataset was downloaded before and the user creates the object with `download=True`. Do you want the default to be `download=True` and _protect_ the user in case he deleted the archives? Otherwise I still don't understand your point.

Reply:
My point is that the user has a single code, written `download=True`, and has run it for the first time and it worked.
Then, he opened the download folder and realized that the zip was still there. He deleted the zip because it take a bunch of space, and then run the same code again, using `download=True` because he hasn't changed it.

Then, he realized that the folder got erased, and it starts to download the tar again!

Reply:
Finally got it. Do you want some logic with a verification file or is a check if the folder is present sufficient?